### DISCUSSION TITLE
Autoplay All Animation

### DISCUSSION THREAD

**Discussion by FaradayNova** [google/model-viewer#2223] [Q&A]
### Description

I have done significant research to find a similar answer, and have found questions close, but either not entirely relevant or not fully pertaining to my exact question

I am exporting from Blender's gltf glb exporter. and have 144 individual objects and 144 nla tracks. 1 track per each
in <modelviewer> i am able to call upon individual animation tracks, and switch between them, including cross fade between animation tracks. This is not what I need

I would like to know if there is a way to play all NLA tracks simultaneously with a single auto play command?

https://gltf-viewer.donmccurdy.com/ had the desired result,(animations - "Play All")

I can't seem to find code or attributes that correlate with the same "Play All" function on https://gltf-viewer.donmccurdy.com/

### Version
- model-viewer: v1.2.1

### Browser Affected

- [x] Chrome
- [x] Edge
- [x] Firefox
- [x] IE
- [x] Safari

### OS

- [x] Android
- [ ] iOS
- [ ] Linux
- [ ] MacOS
- [x] Windows

**Comment by elalish** [comment]
Indeed, model-viewer's API only plays a single animation at a time, cross-fading between. This was simply because animation hasn't been getting a lot of use and this was a simple API. It may be time to expand that now. I'd like to get a bit more information about use cases and API ideas to figure out what would be best to add next. I don't really want to expose all of the three.js animation API, as it's pretty complex.

Also, if you just want to "play all", why not pack them all into a single animation?

**Comment by mithru** [comment]
This is something my team is also exploring. Literally last week, we decided to use three.js over MV directly for this specific use case:  animation of product interacting with a secondary object which is also animated - the secondary object changes while the product stays. Our 3D team found it simpler to animate the objects individually so it can be mixed and matched with other similar 3D objects. 

Right now, we're just using a static model for glb, and for iOS AR we're using a .reality file that includes the animation. We weren't happy with the inconsistency, so switching to threejs.

**Comment by elalish** [comment]
@mithru could you share a link so I can see your use case? Our animation support certainly needs work. Also, PRs are welcome if you have an idea for what would be helpful to expose. Should be fairly easy if you're already familiar with three's API.

**Comment by mithru** [comment]
Got into trouble last time I shared an ongoing project. Sharing something similar instead. Hopefully this explains things better. My apologies if not. 

https://mithru.github.io/AR-test/model-viewer-template/ (please excuse the weird scale here)

In the link above, there are 2 characters in a single loaded GLB. Both of them have separate animations (only one plays on the link I sent, but if you download and open the GLB, you'll see them both animate). The 2 animations could be packaged into a single animation, but that starts to get messy when you want to create multiple combinations of characters with the same choreographed animation. It's not a big lift for smaller sets or one-off animations, but the moment we hit a few combinations, every iteration / design review will have us redo all the GLBs. 

Allowing all animations in the GLB to play at one go would be a quick fix for this particular case. 

What we're building involves a few other things (like mirror materials) that I'm not sure makes sense for MV, so we'll use threejs regardless for this particular project. 

However, I have had a few times in the past where I had to ask my 3D team to send me GLBs with packaged animation. They were small projects so wasn't a big deal to send it back really.

**Comment by elalish** [comment]
@mithru Thanks, that makes sense.

**Reply by kgediya** [reply]
I'm sure this will need to dynamically create multiple animation playback controllers and call play method on them all at once. Currently the architecture is having only one controller I suppose where we can input one animation at a time and control the switching and playback individually

**Reply by mrmaarten** [reply]
It would be nice to have an "play all animations" option in the API.

MV is great for AR projects that serve iOS and Android. iOS just plays all animations in the USDZ file, it would be great to be able to do the same with GLB/GLTF on Android.

**Reply by elalish** [reply]
Is there something that makes it difficult to export the GLB with all the animations joined into one? Does this have an effect on file size?

**Reply by mrmaarten** [reply]
Thanks for the question!

I wasn't aware that you can combine animations into one, so you put me on an interesting/ educational path.

After experimenting I see that is possible to combine animations of different objects also. I even makes the file size smaller.

So thanks for that!

I would still argue that in workflow it is a bit cumbersome: having to have 2 versions of the model. One where all the animations are separate and one where they are all combined. Also: when combining them you have to make sure they are of the same length, they don't loop automatically, so it is extra work to bring all the animations together. 

And when working on the scene you want to keep all animations separate, so when exporting we'd have to combine everything, make sure they are of the same length etc. 

To be clear about the use case: I have a little scene with multiple objects that have an animation. This combination of objects and animations I want to present as an AR experience. 

In this case it is handy to be able to just play all animations, in a loop.

**Reply by elalish** [reply]
Glad it helped! The details you mention are exactly why we've avoided a `playAll` API - pretty quickly what will really be needed is a full animation mixer to deal with things like looping, various lengths, different start times, etc. We may add this eventually, but it feels a little low-level for MV (three.js supports this). Forcing the combination as part of the web-optimization step seems reasonable to keep the authoring separate from the delivery. Still, feedback is welcome!

**Reply by mrmaarten** [reply]
Yeah in my case I actually made my scene smaller by making one animation track.

To elaborate on the use case: you want to present an AR 'object' with animations to the viewer. You want this to work in iOS and Android to make everybody happy. So for iOS you present a USDZ and iOS will handle the rest. On android you present a GLTF and it will be handled.

But if there are multiple animations and you don't know that you can ix everything down to one track, you'd have to dectect the OS and make a separate threejs handler for that just to play all animations. 

So yeah it is either about knowing that all animations can be bounced down to one track, or to have a `playAll` flag in MV.
So it could be added to the documentation, that consolidating animation is an option.

For anybody interested, to mix the animations down in Blender, I used GovieTools for Blender, it also has a buildin webpreviewer for the model, which is very handy. It has great tools for managing animations.
https://blenderartists.org/t/govie-tools-gltf-exporter/1404549
https://github.com/LorenzWieseke/GovieTools

**Comment by jboe182** [comment]
**I have found a solution** Just rename the NLA Strips the same Value like : "1" after exporting into GLB the Animations are "merged in one" without crushing the NLA Timeline in the current project.

**Comment by phovorka** [comment]
TIP: Blender offers by exporting to GLB the option to merge all animation as one... uncheck "Group NLA track" ... that is all.
